rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+
+**Returns:**
+- An optimizer instance with the desired configuration, effectively a clone of the original optimizer (minus its state, i.e. its weight values).
+
+
+### III - Add a get_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.get_weights()
+```
+
+This method is already present on the Model class and every Layer class.
+
+**Returns:**
+- A flat list of Numpy arrays, in deterministic order, where each array represents the value of an internal weight of the optimizer (such as the momentum of a model weight).
+
+
+### IV - Add a set_weights method on every optimizer (only needs a single implementation on the base class):
+
+```python
+optimizer.set_weights(weights)
+```
+
+This method is already present on the Model class and every Layer class. This method is required for Keras compatibility.
+
+**Args:**
+- weights: A flat list of Numpy arrays, in deterministic order, same as returned by get_weights. Note that since the optimizer creates its internal weights to match the set of weights it is trying to optimize, set_weights would only match get_weights when the set of weights being optimized is equivalent. E.g.:
+
+```python
+optimizer = Adadelta()
+_ = optimizer.get_weights()  # returns empty list since the optimizer has no weights at that point
+model.compile(optimizer=optimizer, loss=loss) # Weights are created here
+weights = optimizer.get_weights()  # Returns a list of numpy arrays
+optimizer.set_weights(weights)  # This works!
+
+# This will not work since this optimizer would have a different set weight
+different_model.optimizer.set_weights(weights)
+```
+
+Note: if the optimizer has been called on more than a single set of weights, we should disable `get_weights` and `set_weights` since their meaning would be ambiguous.
+
+
+### V - Make all optimizer hyperparameters accessible via attributes (they currently aren’t retrievable):
+
+```python
+optimizer = Adadelta(learning_rate=0.2)
+optimizer.learning_rate  # returns learning rate tensor
+```
+
+This should generally work for any numerical parameter that can be passed to the constructor.
+
+
+### VI - Make the following work on every optimizer, in both eager and graph modes:
+
+```python
+optimizer = Adadelta(learning_rate=0.2)
+optimizer.learning_rate = 0.1
+```
+
+This should generally work for any numerical parameter that can be passed to the constructor.
+
+In graph mode, this would require 1) creating TF variables for these parameters in the constructor, 2) overriding `__setattr__` to do an assign on the target parameter using the default session.
+
+In eager mode, there are no issues.
+
+
+### VII - Add support for gradient clipping by norm or by value
+
+The following arguments should be supported on all optimizers (it only requires a single shared implementation in the base class):
+
+```python
+Adadelta(clip_norm=0.)
+Adadelta(clip_value=0.)
+```
```

Comment:
I'm against adding this as a universal API because how vaguely the clipping is defined and because there is no standard convention on how clipping should be done.
1. What to clip: do you clip the "gradients w.r.t variables", or do you clip the "updates made to variables"? In most optimizers they are different.
2. How to clip: in tensorflow there are currently `tf.clip_by_{norm, value, global_norm, average_norm}`. Imagine more to come.

For now I can see from above a combination of 2x4=8 methods to do clipping. AFAIK there is no convention on this or no common wisdom on which ones are more preferable. Therefore it's probably not a good idea to add it as an API for all optimizers.

Reply:
A more elegant solution would be to expose finer grained functions as alternative to apply_gradients/,minimize/etc., so that the user may perform any kind of manipulation.

get_updates  (optionally allowing to pass my own gradient values)
apply_updates

Example flows:

  updates = opt.get_updates()
  # clip, log, mask, etc.
  opt.apply_updates(updates)

  grads = opt.get_gradients()  # or tf.gradients
  # clip, etc.
  updates = opt.get_updates(grads=grads)
  opt.apply_updates(updates)

I don't know if this would be at odds with any existing designs.

Reply:
There are already such APIs: `opt.compute_gradients` and `opt.apply_gradients`. Everything is good as long as they are not removed in 2.0.

Reply:
The `compute_gradients`/`apply_gradients` pair only works if you need to manipulate the gradient. If you need to modify the *updates*, you're out of luck.